tribler: 7.4.4 -> 7.10.0

viric opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Updating tribler to 7.10.0-RC2; let's wait until to be 7.10 0 released.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 128145` at ea4d4bf2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built successfully:</summary>
  <ul>
    <li>python38Packages.aiohttp-apispec</li>
    <li>python38Packages.apispec3</li>
    <li>python38Packages.libtorrent-rasterbar-1_2_x</li>
    <li>python38Packages.webargs5</li>
    <li>python39Packages.aiohttp-apispec</li>
    <li>python39Packages.apispec3</li>
    <li>python39Packages.libtorrent-rasterbar-1_2_x</li>
    <li>python39Packages.webargs5</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/aiohttp-apispec/default.nix:20:3:
```
   |
20 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/aiohttp-apispec/default.nix:21:0:
```
   |
21 |     description = "Build and document REST APIs with aiohttp and apispec";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/webargs/5.nix:19:3:
```
   |
19 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/webargs/5.nix:20:0:
```
   |
20 |     description = "Build and document REST APIs with aiohttp and apispec";
   | ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,25 @@
+    sha256 = "0hhlmh3mc3xg68znsxyhypb5k12vg59yf72qkyw6ahg8zy3qfz2m";
+  };
+
+  disabled = pythonOlder "3.0";
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+buildPythonPackage rec {
+  pname = "aiohttp-apispec";
+  version = "2.2.1";
```
```suggestion
  version = "2.2.1";
  disabled = pythonOlder "3.0";
```

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  propagatedBuildInputs = [ aiohttp webargs apispec3 jinja2 ];
+
+  doCheck = false;
```
Please add a comment why tests are disabled and add a pythonImportsCheck.

SuperSandro2000 replied:
I think you just need to add pytestCheckHook to checkInputs and maybe ``tests`` to pytestFlagsArray

viric replied:
I tried the hook thing before, no success. 0 tests.
I will try the pytestFlagsArray.

noughtmare replied:
The tests are not included in the archive file from pypi.

viric replied:
noted

SuperSandro2000 replied:
Can we fetch form source repository instead?

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,25 @@
+  meta = with lib; {
+    description = "Build and document REST APIs with aiohttp and apispec";
+    homepage = "https://github.com/maximdanilchenko/aiohttp-apispec/";
+    license = licenses.mit;
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/apispec/3.nix:
```diff
@@ -0,0 +1,39 @@
+    marshmallow
+    mock
+    pytestCheckHook
+  ];
```
Missing pythonImportsCheck.

viric replied:
done

SuperSandro2000 commented on pkgs/development/python-modules/webargs/5.nix:
```diff
@@ -0,0 +1,24 @@
+  buildInputs = [ pytest webtest ];
+  propagatedBuildInputs = [ marshmallow ];
+
+  doCheck = false;
```
Please add a comment why tests are disabled and add a pythonImportsCheck.

SuperSandro2000 replied:
Try pytestCheckHook.

noughtmare replied:
There are no tests in the pypi archive. Should we get them from GitHub or is that a good enough reason to set `doCheck = false`?

viric replied:
I hope it's a good reason. I added that.

SuperSandro2000 replied:
It is pretty standard to fetch from GitHub instead.

SuperSandro2000 commented on pkgs/development/python-modules/webargs/5.nix:
```diff
@@ -0,0 +1,24 @@
+  meta = with lib; {
+    description = "Build and document REST APIs with aiohttp and apispec";
+    homepage = "https://github.com/maximdanilchenko/aiohttp-apispec/";
+    license = licenses.mit;
```
Missing maintainer.

SuperSandro2000 commented on pkgs/development/python-modules/webargs/5.nix:
```diff
@@ -0,0 +1,24 @@
+    sha256 = "16pjzc265yx579ijz5scffyfd1vsmi87fdcgnzaj2by6w2i445l7";
+  };
+
+  buildInputs = [ pytest webtest ];
```
```suggestion
```

those are checkInputs.

SuperSandro2000 commented on pkgs/development/python-modules/apispec/3.nix:
```diff
@@ -0,0 +1,39 @@
+    description = "A pluggable API specification generator. Currently supports the OpenAPI Specification (f.k.a. the Swagger specification";
+    homepage = "https://github.com/marshmallow-code/apispec";
+    license = licenses.mit;
+    maintainers = [ maintainers.costrouc ];
```
```suggestion
    maintainers = [ maintainers.viric ];
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -285,6 +285,10 @@ in {
   };
 
+  aiohttp-apispec = callPackage ../development/python-modules/aiohttp-apispec {
+    webargs = self.webargs5;
```
That is not allowed.

viric replied:
Can you suggest an alternative?

SuperSandro2000 replied:
It should be done in the project that uses it and not in pythonPackages.

viric replied:
I moved it into aiohttp-apispec, naming the arg webargs5. Fine?

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -9030,6 +9038,10 @@ in {
 
+  webargs = callPackage ../development/python-modules/webargs { };
+
+  webargs5 = callPackage ../development/python-modules/webargs/5.nix { };
```
Please use a pythonOverride instead.

viric replied:
It's not about a different python

noughtmare replied:
You could do it like this:

```diff
diff --git a/pkgs/development/python-modules/aiohttp-apispec/default.nix b/pkgs/development/python-modules/aiohttp-apispec/default.nix
index 31783867456..31fa0d122fc 100644
--- a/pkgs/development/python-modules/aiohttp-apispec/default.nix
+++ b/pkgs/development/python-modules/aiohttp-apispec/default.nix
@@ -2,6 +2,16 @@
 , aiohttp, webargs, apispec3, jinja2
 }:
 
+let
+  webargs5 = webargs.overridePythonAttrs (oldAttrs: rec {
+    version = "5.5.3";
+    src = oldAttrs.src.override {
+      inherit version;
+      sha256 = "16pjzc265yx579ijz5scffyfd1vsmi87fdcgnzaj2by6w2i445l7";
+    };
+    doCheck = false;
+  });
+in
 buildPythonPackage rec {
   pname = "aiohttp-apispec";
   version = "2.2.1";
@@ -12,7 +22,7 @@ buildPythonPackage rec {
     sha256 = "0hhlmh3mc3xg68znsxyhypb5k12vg59yf72qkyw6ahg8zy3qfz2m";
   };
 
-  propagatedBuildInputs = [ aiohttp webargs apispec3 jinja2 ];
+  propagatedBuildInputs = [ aiohttp webargs5 apispec3 jinja2 ];
 
   # No idea how to run the tests
   doCheck = false;
```

Of course the `doCheck = false` is still problematic.

viric replied:
I leave it as is, clarifying that there are no tests in the pypi archive.

SuperSandro2000 replied:
> I leave it as is, clarifying that there are no tests in the pypi archive.

Just do fetchFromGitHub.

> You could do it like this:

It should be overwritten in the application and only exist there.

viric replied:
I tried again to run the webargs 5 tests, and I reached the point until they run but one fails: webapp2 related. Because of a python3 problem with webapp2 that is what makes webapp2 have doCheck=false too.

So do I really have to put aiohttp-apispec and webargs5 within the tribler expression, at the end?

mkg20001 replied:
Another way would be callPackage as overrideAttrs doesn't work too well with python modules

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4068,6 +4074,8 @@ in {
 
   libtorrent-rasterbar = (toPythonModule (pkgs.libtorrent-rasterbar.override { inherit python; })).python;
 
+  libtorrent-rasterbar-1_2_x = (toPythonModule (pkgs.libtorrent-rasterbar-1_2_x.override { inherit python; })).python;
```
Please use a pythonOverride instead.

viric replied:
It's a different libtorrent lib, not a different python.

SuperSandro2000 replied:
I don't think we need this in pythonPackages. We can create it locally in the final program. Same applies to the other new packages.

viric replied:
Ah. Is there any example of this procedure?

SuperSandro2000 replied:
https://github.com/NixOS/nixpkgs/blob/a07c903e121ff73e11637aa479b9f7ce7ce32cf6/pkgs/applications/misc/privacyidea/default.nix#L8

viric replied:
ok I think I did that well with the libtorrent but I don't know how to do the same for the others.

viric commented:
I think I addressed the comments above, @SuperSandro2000. Is it?

noughtmare commented on pkgs/applications/networking/p2p/tribler/default.nix:
```diff
@@ -1,13 +1,20 @@
 { stdenv, fetchurl, python3Packages, makeWrapper
-, enablePlayer ? true, libvlc, qt5, lib }:
+, libtorrent-rasterbar-1_2_x
+, enablePlayer ? true, qt5, lib }:
```
I think this `enablePlayer` option should also be removed.

noughtmare commented on pkgs/applications/networking/p2p/tribler/default.nix:
```diff
@@ -1,13 +1,20 @@
   src = fetchurl {
     url = "https://github.com/Tribler/tribler/releases/download/v${version}/Tribler-v${version}.tar.xz";
-    sha256 = "0hxiyf1k07ngym2p8r1b5mcx1y2crkyz43gi9sgvsvsyijyaff3p";
+    sha256 = "03pav0fi8rxmbwgn2ywd281wx5dwja1w0ds3wgzagpd7nk8iw6pk";
```
I'm getting a different hash:

```
$ nix-prefetch-url https://github.com/Tribler/tribler/releases/download/v7.10.0/Tribler-v7.10.0.tar.xz
[3.2 MiB DL]
path is '/nix/store/y7pd7g4011bxxpv4libwis6l9vkajlja-Tribler-v7.10.0.tar.xz'
1x45z23d1cqf0lai7wg5ki7gi2vba5hqk0swhggzplcjwma4wmh9
```

viric replied:
OH. I also get yours today. Maybe they switched it upstream?

Here is the day when I got the hash: https://github.com/viric/nixpkgs/commit/53fabb2b18ded07515c7932747dcea5fa40ae34f so it is not that I forgot the rc2 hash.

viric replied:
I have the tarballs in the store. It seems I picked their tarball right after the release, because they changed the tarball 12 minutes later.

jonringer commented:
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

 `git rebase -i` is a powerful command which achieves this, I created a small video demonstrating it's use [here](https://www.youtube.com/watch?v=OR8Q-4PLalc). A more indepth text tutorial can be found [here](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)
 
 In short, there should be a commit per package change. Fix up commits should be squashed

mkg20001 commented:
I've fixed things up over here https://github.com/mkg20001/nixpkgs/commit/d252c14e331e75eae794689f1bbecd6cd79ab6cd

This patch would need to be pulled in and rebased against master (can't push to the PRs branch)

SuperSandro2000 commented on pkgs/applications/networking/p2p/tribler/default.nix:
```diff
@@ -1,13 +1,20 @@
+}:
 
+let
+  python = python3Packages.python;
```
This could probably be simplified by using python3 and python3.pkgs.

SuperSandro2000 commented on pkgs/development/python-modules/apispec/3.nix:
```diff
@@ -0,0 +1,43 @@
+    sha256 = "d23ebd5b71e541e031b02a19db10b5e6d5ef8452c552833e3e1afc836b40b1ad";
+  };
+
+  pythonImportsCheck = [
+    "apispec"
+  ];
```
That should be after checkInputs and before meta.

viric commented:
> I've fixed things up over here [mkg20001@d252c14](https://github.com/mkg20001/nixpkgs/commit/d252c14e331e75eae794689f1bbecd6cd79ab6cd)
> 
> This patch would need to be pulled in and rebased against master (can't push to the PRs branch)

I thought I could cherry-pick your fixup patch but it doesn't apply clean. How did you plan that your patch is taken?

viric commented:
I added access to write-access people to add things to the PR

mkg20001 commented:
@viric Propably because I rebased before that. pu -f'd it, now it should be good

viric commented:
I see the squashes erased me from the git history. The requirements from @SuperSandro2000 were beyond my will to integrate this. Good luck @mkg20001.

mkg20001 commented:
@SuperSandro2000 anything against the callPackage solution? Otherwise this should be good

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,38 @@
+  ];
+
+  checkInputs = [
+    pytest
```
Please use pytestCheckHook.

SuperSandro2000 commented on pkgs/development/python-modules/aiohttp-apispec/default.nix:
```diff
@@ -0,0 +1,38 @@
+, aiohttp, webargs, fetchFromGitHub, callPackage, pytest
+}:
+
+let
+  apispec3 = callPackage ../apispec/3.nix {};
+  jinja2 = callPackage ../jinja2/2.nix {};
+in
```
This is sadly not allowed in python-modules because if you have two versions of the same package in one environment python will fall more or less apart. This is a limitation of python, not nix. 

This is no a problem if you move it out of pythonPackages into the package where it is used.

mkg20001 replied:
Ok. It's no longer in top-level/python-modules so I think it should stay in the directory of the module in case multiple programs need the same old version. Recently when I packed something I discovered an old version of a module I needed that way aswell.

SuperSandro2000 replied:
The problem is that if we create a python environment with two different versions of jinja and apispec things will fall apart because Python can't handle that. This needs to be done in tribbler itself but the files can be under python-modules as long as they are not references from there.

viric replied:
So basically aiohttp-apispec has to get out of python-modules and be part of tribler. Right?

SuperSandro2000 replied:
Yes

viric commented:
@SuperSandro2000 you seem to know very clearly what you want to approve this. Can you try to make the change, please? It is blocked since many months.

viric committed:
tribler: 7.4.4 -> 7.10.0

This also adds old versions of apispec and webargs that tribler
requires, and aiohttp-apispec as well.

viric commented:
@SuperSandro2000 moved aiohttpd-apispec to tribler, out of python packages. Anything else?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/i-want-the-python-multi-version-packaging-system-to-be-formally-standardized-by-an-rfc/17537/1
